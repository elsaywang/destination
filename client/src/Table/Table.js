import React, { Component } from 'react';

class Table extends Component {
    render() {
        return <div>{this.props.data}</div>;
    }
}

export default Table;
