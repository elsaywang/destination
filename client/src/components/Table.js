import React from 'react';
import { TableView, TableViewDataSource } from '@react/react-spectrum/TableView';

const columns = {
    keyValuePair: {
        title: 'Key-Value Pair',
        key: 'keyValuePair',
        minWidth: 150,
    },
    keyName: {
        title: 'Key Name',
        key: 'keyName',
        width: 150,
    },
    valueName: {
        title: 'Value Name',
        key: 'valueName',
        width: 150,
    },
    signalType: {
        title: 'Signal Type',
        key: 'signalType',
        width: 140,
    },
    signalSource: {
        title: 'Signal Source',
        key: 'signalSource',
        width: 180,
    },
    totalCounts: {
        title: 'Total Counts',
        key: 'totalCounts',
        width: 100,
    },
    totalEventFires: {
        title: 'Total Event Fires',
        key: 'totalEventFires',
        width: 100,
    },
    percentageChange: {
        title: 'Percentage Change',
        key: 'percentageChange',
        width: 120,
    },
    includedInTraits: {
        title: 'Included In Traits',
        key: 'includedInTraits',
        width: 120,
    },
};

// TODO: Actionable Log Files and General Online Data don't show Signal Source
const basicColumns = [
    columns.keyValuePair,
    columns.signalType,
    columns.signalSource,
    columns.totalCounts,
    columns.percentageChange,
    columns.includedInTraits,
];

const analyticsColumns = [
    columns.keyValuePair,
    columns.keyName,
    columns.valueName,
    columns.totalEventFires,
    columns.percentageChange,
    columns.includedInTraits,
];

const renderCell = (column, data) => <span>{data}</span>;

// Dynamically set the height of the table's container to show up to a certain number of rows.
const getTableHeight = (items, maxRows = 10) => {
    const headHeight = 40;
    const rowHeight = 48;
    const bodyHeight = Math.min(items.length, maxRows) * rowHeight;

    return `${headHeight + bodyHeight}px`;
};

const Table = ({ items }) => {
    const dataSource = new DataSource({ items });

    return (
        <div style={{ height: getTableHeight(items) }}>
            <TableView dataSource={dataSource} renderCell={renderCell} />
        </div>
    );
};

class DataSource extends TableViewDataSource {
    constructor({ items }) {
        super();

        this.items = items;
    }
    // Override
    getColumns() {
        // TODO: Find a way to dynamically return a column group based on a signalType prop.
        // Stuck because the `TableViewDataSource` superclass calls `getColumns()` before props are available.
        return basicColumns;
    }
    // Override
    getNumberOfRows() {
        return this.items.length;
    }
    // Override
    getCell(column, row) {
        return this.items[row][column.key];
    }
    // Override
    sort(column, direction) {
        this.items.sort((a, b) => (a[column.key] < b[column.key] ? -direction : direction));

        this.reloadData();
    }
}

export default Table;
