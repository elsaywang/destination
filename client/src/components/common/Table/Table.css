:local(.tableWrapper) .spectrum-Table-row {
    cursor: auto;
    user-select: text;
}
