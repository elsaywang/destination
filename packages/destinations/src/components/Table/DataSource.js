import ListDataSource from '@react/react-spectrum/ListDataSource';

const createDataSourceSubclass = ({
    getInitialData,
    getMoreData,
    triggerGetInitialData,
    triggerGetMoreData,
    sortData,
}) => {
    return new class DataSource extends ListDataSource {
        createFunctionWithPrototypeChainFallback(prototypeFallbackName, fn) {
            return (...args) => {
                if (fn) {
                    return fn(...args);
                }

                return super[prototypeFallbackName](...args);
            };
        }

        async load(sortDescriptor) {
            const initialData = await getInitialData(sortDescriptor);

            this.rowsLoaded = initialData.length;

            return initialData;
        }

        async loadMore() {
            const dataToAppend = await getMoreData(this.rowsLoaded.length);

            this.rowsLoaded += dataToAppend.length;

            return dataToAppend;
        }

        performLoad(sortDescriptor) {
            return this.createFunctionWithPrototypeChainFallback(
                'performLoad',
                triggerGetInitialData,
            )(sortDescriptor);
        }

        performLoadMore() {
            return this.createFunctionWithPrototypeChainFallback(
                'performLoadMore',
                triggerGetMoreData,
            )();
        }

        performSort(sortDescriptor) {
            return this.createFunctionWithPrototypeChainFallback('performSort', sortData)(
                sortDescriptor,
            );
        }
    }();
};

export default createDataSourceSubclass;
