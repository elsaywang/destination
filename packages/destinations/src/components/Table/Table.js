import React, { Component } from 'react';
import { TableView } from '@react/react-spectrum/TableView';
import ListDataSource from '@react/react-spectrum/ListDataSource';
import IndexPath from '../../utils/indexPath';
import PropTypes from 'prop-types';

class Table extends Component {
    constructor({ reachedEndOfRows, items = [] }) {
        super();

        this.dataSource = new class DataSource extends ListDataSource {
            load() {
                return items;
            }

            performLoadMore() {
                let totalItems = 0;
                for (let i = 0; i < this.getNumberOfSections(); i++) {
                    totalItems += this.getSectionLength(i);
                }

                reachedEndOfRows(totalItems);
            }
        }();
    }

    componentWillReceiveProps(nextProps) {
        const length = this.props.items.length;

        this.dataSource.insertItems(new IndexPath(0, length), nextProps.items.slice(length));
    }

    render() {
        const { dataTest, styles, height, dataSource, items, ...tableViewProps } = this.props;

        return (
            <div data-test={dataTest} className={styles} style={{ height: `${height}px` }}>
                <TableView dataSource={this.dataSource} {...tableViewProps} />
            </div>
        );
    }
}

Table.defaultProps = {
    items: [],
    rowHeight: 30, // chosen randomly, not sure what a sensible default should be
    allowsSelection: false,
    reachedEndOfRows: () => {
        /* no-op*/
    },
};

Table.protoTypes = {
    height: PropTypes.number,
    columns: PropTypes.array,
    rowHeight: PropTypes.number,
    renderCell: PropTypes.func.isRequired,
    onSelectionChange: PropTypes.func,
    dataTest: PropTypes.string,
    allowsSelection: PropTypes.bool,
    styles: PropTypes.string,
    setRefreshDataSource: PropTypes.func,
    reachedEndOfRows: PropTypes.func,
};

export default Table;
