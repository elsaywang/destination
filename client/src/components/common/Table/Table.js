import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { TableView } from '@react/react-spectrum/TableView';
import DataSource from './DataSource';
import {
    defaultRowHeight,
    defaultHeadHeightWithoutSelection,
    defaultMaxRows,
    defaultHeadHeight,
} from '../../../constants/rows';

/**
 * Table component that wraps the react-spectrum table components
 * (`TableView` and `TableViewDataSource`). More easily accepts props
 * for items, columns, table config, and event handlers.
 */
class Table extends Component {
    constructor({ items, columns, sortSearch, onLoadMore, allowsSelection }) {
        super();

        this.state = {
            dataSource: new DataSource({
                items,
                columns,
                sortSearch,
                onLoadMore,
                allowsSelection,
            }),
        };
    }

    /**
     * Dynamically set the height of the table's container to show up to a
     * certain number of rows.
     */
    getTableHeight({
        items,
        rowHeight = defaultRowHeight,
        allowsSelection = true,
        maxRows = defaultMaxRows,
    }) {
        const headHeight = allowsSelection ? defaultHeadHeight : defaultHeadHeightWithoutSelection;
        const bodyHeight = Math.min(items.length, maxRows) * rowHeight;

        return `${headHeight + bodyHeight}px`;
    }

    shouldComponentUpdate(nextProps) {
        if (!this.props.selectedRowIndexes || !nextProps.selectedRowIndexes) {
            return true;
        }
        if (this.props.isMaxSelectedRowsReached || nextProps.isMaxSelectedRowsReached) {
            return false;
        }
        const getSelectedRowIndexes = ({ selectedRowIndexes }) =>
            JSON.stringify(selectedRowIndexes.sort());

        return getSelectedRowIndexes(nextProps) === getSelectedRowIndexes(this.props);
    }

    componentDidUpdate() {
        const { items, columns } = this.props;

        this.state.dataSource.items = items;
        this.state.dataSource.columns = columns;
        this.state.dataSource.reloadData();
    }

    render() {
        const {
            items,
            rowHeight,
            renderCell,
            onSelectionChange,
            dataTest,
            allowsSelection,
        } = this.props;

        const height = this.getTableHeight({ items, rowHeight, allowsSelection });

        return (
            <div data-test={dataTest} className="table-wrapper" style={{ height }}>
                <TableView
                    dataSource={this.state.dataSource}
                    renderCell={renderCell}
                    rowHeight={rowHeight}
                    onSelectionChange={onSelectionChange}
                    allowsSelection={allowsSelection}
                />
            </div>
        );
    }
}

Table.propTypes = {
    items: PropTypes.array.isRequired,
    columns: PropTypes.array.isRequired,
    renderCell: PropTypes.func.isRequired,
    rowHeight: PropTypes.number,
    selectedRowIndexes: PropTypes.array,
    isMaxSelectedRowsReached: PropTypes.bool,
    onSelectionChange: PropTypes.func,
    sortSearch: PropTypes.func,
    dataTest: PropTypes.string,
    allowsSelection: PropTypes.bool,
};

export default Table;
