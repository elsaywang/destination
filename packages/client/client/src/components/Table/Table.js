import PropTypes from "prop-types";
import React, { Component } from "react";
import classNames from "classnames";
import { TableView } from "@react/react-spectrum/TableView";
import DataSource from "./DataSource";
import {
	defaultRowHeight,
	defaultMaxRows,
	defaultHeadHeight
} from "../../constants/rows";
import styles from "./Table.css";

/**
 * Table component that wraps the react-spectrum table components
 * (`TableView` and `TableViewDataSource`). More easily accepts props
 * for items, columns, table config, and event handlers.
 */
class Table extends Component {
	constructor({ items, onLoadMore }) {
		super();
		this.dataSource = new DataSource({ items, onLoadMore });
	}

	/**
	 * Dynamically set the height of the table's container to show up to a
	 * certain number of rows.
	 */
	getTableHeight({
		items,
		rowHeight = defaultRowHeight,
		maxRows = defaultMaxRows
	}) {
		const bodyHeight = Math.min(items.length, maxRows) * rowHeight;

		return `${defaultHeadHeight + bodyHeight}px`;
	}

	render() {
		const {
			items,
			columns,
			rowHeight,
			renderCell,
			onSelectionChange,
			dataTest,
			allowsSelection
		} = this.props;

		const height = this.getTableHeight({ items, rowHeight });
		return (
			<div
				data-test={dataTest}
				className={classNames("table-wrapper", styles.tableWrapper)}
				style={{ height }}>
				<TableView
					dataSource={this.dataSource}
					columns={columns}
					renderCell={renderCell}
					rowHeight={rowHeight}
					onSelectionChange={onSelectionChange}
					allowsSelection={allowsSelection}
				/>
			</div>
		);
	}
}

Table.propTypes = {
	items: PropTypes.array.isRequired,
	columns: PropTypes.array.isRequired,
	renderCell: PropTypes.func.isRequired,
	onLoadMore: PropTypes.func,
	rowHeight: PropTypes.number,
	selectedRowIndexes: PropTypes.array,
	isMaxSelectedRowsReached: PropTypes.bool,
	onSelectionChange: PropTypes.func,
	sortSearch: PropTypes.func,
	dataTest: PropTypes.string,
	allowsSelection: PropTypes.bool
};

export default Table;
