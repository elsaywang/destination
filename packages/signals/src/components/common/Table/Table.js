import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classNames from 'classnames';
import { TableView } from '@react/react-spectrum/TableView';
import DataSource from './DataSource';
import { defaultRowHeight, defaultMaxRows, defaultHeadHeight } from '../../../constants/rows';
import styles from './Table.css';

/**
 * Table component that wraps the react-spectrum table components
 * (`TableView` and `TableViewDataSource`). More easily accepts props
 * for items, columns, table config, and event handlers.
 */
class Table extends Component {
    constructor({ items, columns, sortSearch, onLoadMore, allowsSelection }) {
        super();

        this.state = {
            dataSource: new DataSource({
                items,
                columns,
                sortSearch,
                onLoadMore,
                allowsSelection,
            }),
        };
    }

    /**
     * Dynamically set the height of the table's container to show up to a
     * certain number of rows.
     */
    getTableHeight({ items, rowHeight = defaultRowHeight, maxRows = defaultMaxRows }) {
        const bodyHeight = Math.min(items.length, maxRows) * rowHeight;

        return `${defaultHeadHeight + bodyHeight}px`;
    }

    shouldComponentUpdate(nextProps) {
        // Skip render optimizations for Dashboard tables, which don't have `selectedRowIndexes`.
        if (!this.props.selectedRowIndexes || !nextProps.selectedRowIndexes) {
            return true;
        }
        if (this.props.isMaxSelectedRowsReached || nextProps.isMaxSelectedRowsReached) {
            return false;
        }
        const getSelectedRowIndexes = ({ selectedRowIndexes }) =>
            JSON.stringify(selectedRowIndexes.sort());

        return getSelectedRowIndexes(nextProps) === getSelectedRowIndexes(this.props);
    }

    componentDidUpdate() {
        const { items, columns } = this.props;

        this.state.dataSource.items = items;
        this.state.dataSource.columns = columns;
        this.state.dataSource.reloadData();
    }

    render() {
        const {
            items,
            rowHeight,
            renderCell,
            onSelectionChange,
            dataTest,
            allowsSelection,
        } = this.props;

        const height = this.getTableHeight({ items, rowHeight });

        return (
            <div
                data-test={dataTest}
                className={classNames('table-wrapper', styles.tableWrapper)}
                style={{ height }}>
                <TableView
                    dataSource={this.state.dataSource}
                    renderCell={renderCell}
                    rowHeight={rowHeight}
                    onSelectionChange={onSelectionChange}
                    allowsSelection={allowsSelection}
                />
            </div>
        );
    }
}

Table.propTypes = {
    items: PropTypes.array.isRequired,
    columns: PropTypes.array.isRequired,
    renderCell: PropTypes.func.isRequired,
    rowHeight: PropTypes.number,
    selectedRowIndexes: PropTypes.array,
    isMaxSelectedRowsReached: PropTypes.bool,
    onSelectionChange: PropTypes.func,
    sortSearch: PropTypes.func,
    dataTest: PropTypes.string,
    allowsSelection: PropTypes.bool,
};

export default Table;
