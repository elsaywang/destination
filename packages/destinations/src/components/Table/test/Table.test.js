import React from 'react';
import { shallow } from 'enzyme';
import Table from '../Table';
import { TableView } from '@react/react-spectrum/TableView';

describe('<Table/> component', () => {
    describe('rendering', () => {
        const sort = jest.fn();

        const props = {
            renderCell: (column, data) => {
                return <span>{data[column.key]}</span>;
            },
            columns: [
                {
                    title: 'ID',
                    key: 'id',
                    width: 50,
                    sortable: true,
                },
                {
                    title: 'Name',
                    key: 'name',
                    width: 50,
                    sortable: true,
                },
                {
                    title: 'Icons',
                    key: 'icons',
                    width: 50,
                    sortable: false,
                },
            ],
            onSortChange: sort,
        };

        const wrapper = shallow(<Table {...props} />);

        it('matches snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('renders <TableView>', () => {
            expect(wrapper.find(TableView)).toBeTruthy();
        });
    });
});
