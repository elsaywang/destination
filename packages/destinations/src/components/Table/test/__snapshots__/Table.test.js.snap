// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table/> component rendering matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Table
    allowsSelection={false}
    columns={
        Array [
            Object {
              "key": "id",
              "sortable": true,
              "title": "ID",
              "width": 50,
            },
            Object {
              "key": "name",
              "sortable": true,
              "title": "Name",
              "width": 50,
            },
            Object {
              "key": "icons",
              "sortable": false,
              "title": "Icons",
              "width": 50,
            },
          ]
    }
    items={Array []}
    onSortChange={[Function]}
    reachedEndOfRows={[Function]}
    renderCell={[Function]}
    rowHeight={30}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <TableView
        acceptsDrops={false}
        allowsMultipleSelection={true}
        allowsSelection={false}
        canDragItems={false}
        columns={
                Array [
                        Object {
                          "key": "id",
                          "sortable": true,
                          "title": "ID",
                          "width": 50,
                        },
                        Object {
                          "key": "name",
                          "sortable": true,
                          "title": "Name",
                          "width": 50,
                        },
                        Object {
                          "key": "icons",
                          "sortable": false,
                          "title": "Icons",
                          "width": 50,
                        },
                      ]
        }
        dataSource={
                DataSource {
                        "_events": Object {},
                        "_eventsCount": 0,
                        "_maxListeners": undefined,
                        "sections": Array [],
                      }
        }
        dropPosition="between"
        onSortChange={[Function]}
        reachedEndOfRows={[Function]}
        renderCell={[Function]}
        rowHeight={30}
/>,
      "className": undefined,
      "data-test": undefined,
      "style": Object {
        "height": "undefinedpx",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "acceptsDrops": false,
        "allowsMultipleSelection": true,
        "allowsSelection": false,
        "canDragItems": false,
        "columns": Array [
          Object {
            "key": "id",
            "sortable": true,
            "title": "ID",
            "width": 50,
          },
          Object {
            "key": "name",
            "sortable": true,
            "title": "Name",
            "width": 50,
          },
          Object {
            "key": "icons",
            "sortable": false,
            "title": "Icons",
            "width": 50,
          },
        ],
        "dataSource": DataSource {
          "_events": Object {},
          "_eventsCount": 0,
          "_maxListeners": undefined,
          "sections": Array [],
        },
        "dropPosition": "between",
        "onSortChange": [Function],
        "reachedEndOfRows": [Function],
        "renderCell": [Function],
        "rowHeight": 30,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <TableView
          acceptsDrops={false}
          allowsMultipleSelection={true}
          allowsSelection={false}
          canDragItems={false}
          columns={
                    Array [
                              Object {
                                "key": "id",
                                "sortable": true,
                                "title": "ID",
                                "width": 50,
                              },
                              Object {
                                "key": "name",
                                "sortable": true,
                                "title": "Name",
                                "width": 50,
                              },
                              Object {
                                "key": "icons",
                                "sortable": false,
                                "title": "Icons",
                                "width": 50,
                              },
                            ]
          }
          dataSource={
                    DataSource {
                              "_events": Object {},
                              "_eventsCount": 0,
                              "_maxListeners": undefined,
                              "sections": Array [],
                            }
          }
          dropPosition="between"
          onSortChange={[Function]}
          reachedEndOfRows={[Function]}
          renderCell={[Function]}
          rowHeight={30}
/>,
        "className": undefined,
        "data-test": undefined,
        "style": Object {
          "height": "undefinedpx",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "acceptsDrops": false,
          "allowsMultipleSelection": true,
          "allowsSelection": false,
          "canDragItems": false,
          "columns": Array [
            Object {
              "key": "id",
              "sortable": true,
              "title": "ID",
              "width": 50,
            },
            Object {
              "key": "name",
              "sortable": true,
              "title": "Name",
              "width": 50,
            },
            Object {
              "key": "icons",
              "sortable": false,
              "title": "Icons",
              "width": 50,
            },
          ],
          "dataSource": DataSource {
            "_events": Object {},
            "_eventsCount": 0,
            "_maxListeners": undefined,
            "sections": Array [],
          },
          "dropPosition": "between",
          "onSortChange": [Function],
          "reachedEndOfRows": [Function],
          "renderCell": [Function],
          "rowHeight": 30,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
